# Exercice 1

## Contexte et objectifs

L'objectif de ce premier défi est de se familiariser avec la conception de bases de données, une compétence **cruciale** pour rendre votre base de donnée maintenable, flexible et efficace. Pour créer nos bases de données, nous utiliserons le [SQL Designer](https://sql.toad.cz/), un outil que nous utiliserons plusieurs fois au cours de la semaine.

## Specs

### Design de notre première table

Commençons par créer une table pour stocker les utilisateurs. La table doit contenir les colonnes suivantes
- First name
- Last name
- Email

Utilise le [SQL Designer](https://sql.toad.cz/) pour ajouter la table `users` avec les colonnes spécifiées ci-dessus.

NOTE : Rappelle-toi que chaque table a besoin d'une colonne `id` !

### Design de la base de données de surveys

Continuons à partir de notre table `users` pour construire une base de données `surveys`. Il existe de nombreuses façons de construire une base de données d'enquête et de sondages, mais commençons par construire un système de base avec des utilisateurs, des questions, des réponses et des enquêtes.

Voici les exigences de notre système :
- L'application gère plusieurs `users` (que nous avons déjà).
- Un utilisateur peut créer plusieurs `surveys`, mais une enquête n'est créée que par un seul utilisateur.
- Une enquête comporte plusieurs `questions`.
- Une question a plusieurs `answers` possibles.
- Lorsqu'un utilisateur répond à une question, il ne peut choisir qu'une seule réponse. Celles-ci sont stockées en tant que `user_answers`.

### Design du schéma

Utilisons à nouveau le [SQL Designer](https://sql.toad.cz/) pour concevoir un schéma de base de données pour une application d'enquête qui répond aux exigences ci-dessus.

Insère le schéma de la base de donnée sous forme de capture d'écran dans la cellule suivante.

## Points clés de l'apprentissage

- Sais-tu ce qu'est un schéma ?
- Quelle est la relation entre les tables ?
- Peux-tu dessiner le schéma de la base de données derrière Facebook ? Airbnb ? Prends une feuille et essaye de deviner les différentes tables !

![students db table](./images/Ex1.png "This is my answer")

# Excercice 2

Spécifications
L’objectif de cet exercice est d’explorer la base de données Jukebox et de comprendre son schéma. Réponds aux questions suivantes :

- Quel est le schéma de la base de données ? (quelles sont les tables et les relations entre les tables)
- Utilise l’outil SQL Design pour dessiner le schéma de cette base de données.
- Quels sont les noms des colonnes pour chaque table ?

Utilise à nouveau le [SQL Designer](https://sql.toad.cz/) pour concevoir le schéma de base de la base de données.

### Petit coup de main :

In [11]:
import sqlite3
import pandas as pd

# Connect to the SQLite database
conn = sqlite3.connect('./data/jukebox.sqlite')

# Query the sqlite_master table to get the schema of all tables
query = '''
SELECT name
FROM sqlite_master
WHERE type IN ('table', 'view') AND name != 'sqlite_stat1'
'''

# Create a pandas DataFrame from the SQL query
df = pd.read_sql_query(query, conn)

# Display the DataFrame
df

name
0       albums
1      artists
2       genres
3  media_types
4       tracks

In [12]:
# Create a pandas DataFrame from a SQL query
sql_query = '''

SELECT * FROM tracks

'''

df = pd.read_sql_query(sql_query, conn)

df

id  album_id  media_type_id  genre_id  \
0        1         1              1         1   
1        2         2              2         1   
2        3         3              2         1   
3        4         3              2         1   
4        5         3              2         1   
...    ...       ...            ...       ...   
3498  3499       343              2        24   
3499  3500       344              2        24   
3500  3501       345              2        24   
3501  3502       346              2        24   
3502  3503       347              2        10   

                                                   name  \
0               For Those About To Rock (We Salute You)   
1                                     Balls to the Wall   
2                                       Fast As a Shark   
3                                     Restless and Wild   
4                                  Princess of the Dawn   
...                                                 ...   
3498  Pini Di Roma (Pinien Von Rom) \ I Pini Della V...   
3499  String Quartet No. 12 in C Minor, D. 703 "Quar...   
3500               L'orfeo, Act 3, Sinfonia (Orchestra)   
3501  Quintet for Horn, Violin, 2 Violas, and Cello ...   
3502                                      Koyaanisqatsi   

                                               composer  milliseconds  \
0             Angus Young, Malcolm Young, Brian Johnson        343719   
1                                                  None        342562   
2     F. Baltes, S. Kaufman, U. Dirkscneider & W. Ho...        230619   
3     F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. D...        252051   
4                            Deaffy & R.A. Smith-Diesel        375418   
...                                                 ...           ...   
3498                                               None        286741   
3499                                     Franz Schubert        139200   
3500                                 Claudio Monteverdi         66639   
3501                            Wolfgang Amadeus Mozart        221331   
3502                                       Philip Glass        206005   

         bytes  unit_price  
0     11170334        0.99  
1      5510424        0.99  
2      3990994        0.99  
3      4331779        0.99  
4      6290521        0.99  
...        ...         ...  
3498   4718950        0.99  
3499   2283131        0.99  
3500   1189062        0.99  
3501   3665114        0.99  
3502   3305164        0.99  

[3503 rows x 9 columns]

![students db table](./images/Ex2.png "This is my answer")

# Exercice 3

## Contexte et objectifs

L’objectif de cet exercice est de communiquer avec la base de données **depuis Python**.

Dans ce challenge, tu vas travailler exclusivement dans le jupyter notebook, dans lequel tu devras compléter plusieurs cellules pour rassembler des données précises provenant de la base de données.

## Spécifications

👉 **IMPORTANT** : Inspire toi du code python inséré ci-dessus pour tester ta requête SQL, pas besoin de modifier le code en Python :

```python
sql_query = '''

SELECT COUNT *
FROM artists


'''

df = pd.read_sql_query(sql_query, conn)

df
```

## Specs

Il y a six requêtes SQL à compléter :

- Combien de lignes la table `artists` contient-elle ? : 275 lignes
SELECT COUNT (*)
FROM artists

- Combien de lignes chaque table contient-elle ?
albums : 347 lignes 
artist : 275 lignes
genres : 25 lignes
media_types : 5 lignes
tracks : 3503 lignes

- Retourne la liste de tous les artistes classés par nom (par ordre alphabétique). **Indice :** utilise le filtre SQL `ORDER BY`.
SELECT name
FROM artists
ORDER BY name ASC

- Trouve toutes les chansons d’amour (à savoir les chansons dont le titre contient le mot "love" **n'importe où**). **Indice :** utilise les mots-clés SQL `WHERE` et `LIKE`.
SELECT *
FROM tracks
WHERE name LIKE "%love%"

=> C'est la liste de 114 chansons qui contient le mot "love"

- Retourne toutes les chansons dont la durée est supérieure à 45 minutes et trie-les. **Indice :** tu peux utiliser l’opérateur de comparaison `>` dans SQL.
SELECT *
FROM tracks
WHERE milliseconds > 2700000

=> C'est la liste de 31 chansons qui a le millesecond supérieur à 2700000 (45 minutes)

- Retourne la liste de chaque artiste avec leur nombre d'albums. **Indice :** utilise le mot-clé SQL `JOIN`, ainsi que `GROUP BY` et `ORDER BY`.
SELECT artists.name, count(albums.id)
FROM artists
JOIN albums ON albums.artist_id = artists.id
GROUP BY artists.id
ORDER BY artists.name ASC

=> c'est la liste de 203 artistes avec leur nombre d'albums

## Conseils

Les requêtes SQL ont tendance à être longues, notamment si tu commences à utiliser `WHERE` ou `JOIN`. En Python, tu peux utiliser le multiligne pour écrire tes requêtes à **plusieurs lignes** n'oublie pas d'utiliser les ''' :

```python
# Trouve les 3 premiers artistes avec la lettre `Z` dans leur nom.
query = '''
 SELECT * FROM artists
 WHERE name LIKE "%Z%"
 ORDER BY name
 LIMIT 3
'''
df = pd.read_sql_query(query, conn)
```


## Ressources

- [Commandes SQL](https://www.codecademy.com/article/sql-commands)
- [Cours en 🇫🇷 sur `SELECT`](http://sqlpro.developpez.com/cours/sqlaz/select/#L3.4)
- [Documentation SQL](https://sql.sh/)

## Combien de lignes la table `artists` contient-elle ?

In [5]:
import sqlite3
import pandas as pd

# Connect to the SQLite database
conn = sqlite3.connect('./data/jukebox.sqlite')

# Query the sqlite_master table to get the schema of all tables
query = '''
SELECT COUNT (*) as nbr_lignes_artists
FROM artists

'''

# Create a pandas DataFrame from the SQL query
df = pd.read_sql_query(query, conn)

# Display the DataFrame
df

nbr_lignes_artists
0                 275

## Combien de lignes chaque table contient-elle ?

In [4]:
# Query the sqlite_master table to get the schema of all tables
query = '''
SELECT COUNT (*) as nbr_lignes_albums
FROM albums

'''

# Create a pandas DataFrame from the SQL query
df = pd.read_sql_query(query, conn)

# Display the DataFrame
df

nbr_lignes_albums
0                347

In [6]:
# Query the sqlite_master table to get the schema of all tables
query = '''
SELECT COUNT (*) as nbr_lignes_genres
FROM genres

'''

# Create a pandas DataFrame from the SQL query
df = pd.read_sql_query(query, conn)

# Display the DataFrame
df

nbr_lignes_genres
0                 25

In [7]:
# Query the sqlite_master table to get the schema of all tables
query = '''
SELECT COUNT (*) as nbr_lignes_media_types
FROM media_types

'''

# Create a pandas DataFrame from the SQL query
df = pd.read_sql_query(query, conn)

# Display the DataFrame
df

nbr_lignes_media_types
0                       5

In [8]:
# Query the sqlite_master table to get the schema of all tables
query = '''
SELECT COUNT (*) as nbr_lignes_tracks
FROM tracks

'''

# Create a pandas DataFrame from the SQL query
df = pd.read_sql_query(query, conn)

# Display the DataFrame
df

nbr_lignes_tracks
0               3503

## Retourne la liste de tous les artistes classés par nom (par ordre alphabétique).

In [9]:
# Query the sqlite_master table to get the schema of all tables
query = '''

SELECT name
FROM artists
ORDER BY name ASC

'''

# Create a pandas DataFrame from the SQL query
df = pd.read_sql_query(query, conn)

# Display the DataFrame
df


name
0                                         A Cor Do Som
1                                                AC/DC
2            Aaron Copland & London Symphony Orchestra
3                                       Aaron Goldberg
4    Academy of St. Martin in the Fields & Sir Nevi...
..                                                 ...
270                                                Xis
271                                     Yehudi Menuhin
272                                           Yo-Yo Ma
273                                     Youssou N'Dour
274                                     Zeca Pagodinho

[275 rows x 1 columns]

## Trouve toutes les chansons d’amour (à savoir les chansons dont le titre contient le mot "love" **n'importe où**). 

In [10]:
# Query the sqlite_master table to get the schema of all tables
query = '''

SELECT *
FROM tracks
WHERE name LIKE "%love%"

'''

# Create a pandas DataFrame from the SQL query
df = pd.read_sql_query(query, conn)

# Display the DataFrame
df



id  album_id  media_type_id  genre_id  \
0      24         5              1         1   
1      56         7              1         1   
2     195        20              1         6   
3     335        29              1         9   
4     341        30              1         1   
..    ...       ...            ...       ...   
109  3355       265              5         1   
110  3377       270              2        23   
111  3460       321              2        14   
112  3470       322              2         9   
113  3471       322              2         9   

                                             name  \
0                             Love In An Elevator   
1                                Love, Hate, Love   
2                            Let Me Love You Baby   
3                                         My Love   
4    The Girl I Love She Got Long Black Wavy Hair   
..                                            ...   
109                                    Love Comes   
110                         Arms Around Your Love   
111                         Love Is a Losing Game   
112                         I Heard Love Is Blind   
113        (There Is) No Greater Love (Teo Licks)   

                                              composer  milliseconds  \
0                              Steven Tyler, Joe Perry        321828   
1                         Jerry Cantrell, Layne Staley        387134   
2                                         Willie Dixon        175386   
3                                     Jauperi/Zeu Góes        203493   
4    Jimmy Page/John Bonham/John Estes/John Paul Jo...        183327   
..                                                 ...           ...   
109  Darius "Take One" Minwalla/Jon Auer/Ken String...        199923   
110                                      Chris Cornell        214016   
111                                               None        154386   
112                                               None        129666   
113                          Isham Jones & Marty Symes        167933   

        bytes  unit_price  
0    10552051        0.99  
1    12575396        0.99  
2     5716994        0.99  
3     6772813        0.99  
4     5995686        0.99  
..        ...         ...  
109   3240609        0.99  
110   3516224        0.99  
111   2509409        0.99  
112   2190831        0.99  
113   2773507        0.99  

[114 rows x 9 columns]

## Retourne toutes les chansons dont la durée est supérieure à 45 minutes et trie-les. 

In [11]:
# Query the sqlite_master table to get the schema of all tables
query = '''

SELECT *
FROM tracks
WHERE milliseconds > 2700000

'''

# Create a pandas DataFrame from the SQL query
df = pd.read_sql_query(query, conn)

# Display the DataFrame
df

id  album_id  media_type_id  genre_id  \
0   2820       227              3        19   
1   2826       227              3        18   
2   2838       227              3        20   
3   2902       231              3        21   
4   2910       231              3        19   
5   2918       231              3        19   
6   2920       231              3        19   
7   3224       229              3        21   
8   3226       253              3        20   
9   3227       253              3        20   
10  3228       253              3        20   
11  3229       253              3        20   
12  3230       253              3        20   
13  3231       253              3        20   
14  3232       253              3        20   
15  3233       253              3        20   
16  3234       253              3        20   
17  3235       253              3        20   
18  3236       253              3        20   
19  3237       253              3        20   
20  3238       253              3        20   
21  3239       253              3        20   
22  3240       253              3        20   
23  3241       253              3        20   
24  3242       253              3        20   
25  3243       253              3        20   
26  3244       253              3        20   
27  3245       253              3        20   
28  3246       253              3        20   
29  3247       253              3        20   
30  3248       253              3        20   
31  3249       253              3        20   

                                 name composer  milliseconds       bytes  \
0              Occupation / Precipice     None       5286953  1054423946   
1                                Hero     None       2713755   506896959   
2                   Crossroads, Pt. 2     None       2869953   497335706   
3                     Maternity Leave     None       2780416   555244214   
4                                Dave     None       2825166   574325829   
5                                 "?"     None       2782333   528227089   
6                       Three Minutes     None       2763666   531556853   
7             Through a Looking Glass     None       5088838  1059546140   
8         Battlestar Galactica, Pt. 1     None       2952702   541359437   
9         Battlestar Galactica, Pt. 2     None       2956081   521387924   
10        Battlestar Galactica, Pt. 3     None       2927802   554509033   
11     Lost Planet of the Gods, Pt. 1     None       2922547   537812711   
12     Lost Planet of the Gods, Pt. 2     None       2914664   534343985   
13                   The Lost Warrior     None       2920045   558872190   
14                    The Long Patrol     None       2925008   513122217   
15  The Gun On Ice Planet Zero, Pt. 1     None       2907615   540980196   
16  The Gun On Ice Planet Zero, Pt. 2     None       2924341   546542281   
17           The Magnificent Warriors     None       2924716   570152232   
18                    The Young Lords     None       2863571   587051735   
19           The Living Legend, Pt. 1     None       2924507   503641007   
20           The Living Legend, Pt. 2     None       2923298   515632754   
21                      Fire In Space     None       2926593   536784757   
22             War of the Gods, Pt. 1     None       2922630   505761343   
23             War of the Gods, Pt. 2     None       2923381   487899692   
24            The Man With Nine Lives     None       2956998   577829804   
25          Murder On the Rising Star     None       2935894   551759986   
26        Greetings from Earth, Pt. 1     None       2960293   536824558   
27        Greetings from Earth, Pt. 2     None       2903778   527842860   
28                    Baltar's Escape     None       2922088   525564224   
29                Experiment In Terra     None       2923548   547982556   
30                  Take the Celestra     None       2927677   512381289   
31                    

## Retourne la liste de chaque artiste avec leur nombre d'albums

In [12]:
# Query the sqlite_master table to get the schema of all tables
query = '''

SELECT artists.name, count(albums.id)
FROM artists
JOIN albums ON albums.artist_id = artists.id
GROUP BY artists.id
ORDER BY artists.name ASC

'''

# Create a pandas DataFrame from the SQL query
df = pd.read_sql_query(query, conn)

# Display the DataFrame
df

name  count(albums.id)
0                                                AC/DC                 2
1            Aaron Copland & London Symphony Orchestra                 1
2                                       Aaron Goldberg                 1
3    Academy of St. Martin in the Fields & Sir Nevi...                 1
4    Academy of St. Martin in the Fields Chamber En...                 1
..                                                 ...               ...
199                                 Vinícius De Moraes                 1
200                                     Wilhelm Kempff                 1
201                                     Yehudi Menuhin                 1
202                                           Yo-Yo Ma                 1
203                                     Zeca Pagodinho                 1

[204 rows x 2 columns]

# Exercice 4 (join queries)

## Contexte et objectifs

Le moment est venu de passer à quelque chose de plus complexe. On va utiliser les requêtes `JOIN` pour lire les données de plusieurs tables. Pour devenir instantanément un pro des requêtes `JOIN`, [lis ceci](http://stackoverflow.com/questions/17946221/sql-join-and-different-types-of-joins). L’image est très utile. Tu peux aussi [lire ceci](http://sql.sh/cours/jointures).

## Spécifications

Reutilise le code python exactement comme dans l’exercice précédent pour pouvoir interroger la base de donnée.

### Voici les requêtes : 

- La liste des **morceaux avec leur album et artiste**.
- Pour le genre musical 'Jazz', retourner le **nombre de morceaux** et la **durée moyenne des chansons**. Indice : tu peux utiliser la fonction SQL `AVG` pour _average_
- Lister les informations des morceaux et de l'album pour les morceaux dont les noms contiennent un mot-clé donné. Choisis le mot-clé. Exemple : `love`
- Lister les top 5 artistes de Rock ayant le plus de morceaux.


### La liste des **morceaux avec leur album et artiste**.

In [1]:
import sqlite3
import pandas as pd

# Connect to the SQLite database
conn = sqlite3.connect('./data/jukebox.sqlite')

# Query the sqlite_master table to get the schema of all tables
query = '''
SELECT artists.name as artists, tracks.name as chansons, albums.title as albums
FROM tracks
JOIN albums ON tracks.album_id = albums.id
JOIN artists ON albums.artist_id = artists.id
'''

# Create a pandas DataFrame from the SQL query
df = pd.read_sql_query(query, conn)

# Display the DataFrame
df


artists  \
0                                                 AC/DC   
1                                                Accept   
2                                                Accept   
3                                                Accept   
4                                                Accept   
...                                                 ...   
3498                                     Eugene Ormandy   
3499                             Emerson String Quartet   
3500  C. Monteverdi, Nigel Rogers - Chiaroscuro; Lon...   
3501                                      Nash Ensemble   
3502                              Philip Glass Ensemble   

                                               chansons  \
0               For Those About To Rock (We Salute You)   
1                                     Balls to the Wall   
2                                       Fast As a Shark   
3                                     Restless and Wild   
4                                  Princess of the Dawn   
...                                                 ...   
3498  Pini Di Roma (Pinien Von Rom) \ I Pini Della V...   
3499  String Quartet No. 12 in C Minor, D. 703 "Quar...   
3500               L'orfeo, Act 3, Sinfonia (Orchestra)   
3501  Quintet for Horn, Violin, 2 Violas, and Cello ...   
3502                                      Koyaanisqatsi   

                                                 albums  
0                 For Those About To Rock We Salute You  
1                                     Balls to the Wall  
2                                     Restless and Wild  
3                                     Restless and Wild  
4                                     Restless and Wild  
...                                                 ...  
3498                             Respighi:Pines of Rome  
3499  Schubert: The Late String Quartets & String Qu...  
3500                                Monteverdi: L'Orfeo  
3501                              Mozart: Chamber Music  
3502  Koyaanisqatsi (Soundtrack from the Motion Pict...  

[3503 rows x 3 columns]

### Pour le genre musical 'Jazz', retourner le **nombre de morceaux** et la **durée moyenne des chansons**.

In [80]:
import sqlite3
import pandas as pd

# Connect to the SQLite database
conn = sqlite3.connect('./data/jukebox.sqlite')

# Query the sqlite_master table to get the schema of all tables
query = '''
SELECT count(*), AVG(tracks.milliseconds/60000)
FROM tracks
JOIN albums ON tracks.album_id = albums.id
JOIN genres ON tracks.genre_id = genres.id
WHERE genres.name LIKE "%Jazz%"
'''

# Create a pandas DataFrame from the SQL query
df = pd.read_sql_query(query, conn)

# Display the DataFrame
df

count(*)  AVG(tracks.milliseconds/60000)
0       130                             4.4

### Lister les informations des morceaux et de l'album pour les morceaux dont les noms contiennent un mot-clé donné. 

In [85]:
import sqlite3
import pandas as pd

# Connect to the SQLite database
conn = sqlite3.connect('./data/jukebox.sqlite')

# Query the sqlite_master table to get the schema of all tables
query = '''
SELECT *
FROM tracks
JOIN albums ON tracks.album_id = albums.id
WHERE tracks.name LIKE "%love%"
'''

# Create a pandas DataFrame from the SQL query
df = pd.read_sql_query(query, conn)

# Display the DataFrame
df

id  album_id  media_type_id  genre_id  \
0      24         5              1         1   
1      56         7              1         1   
2     195        20              1         6   
3     335        29              1         9   
4     341        30              1         1   
..    ...       ...            ...       ...   
109  3355       265              5         1   
110  3377       270              2        23   
111  3460       321              2        14   
112  3470       322              2         9   
113  3471       322              2         9   

                                             name  \
0                             Love In An Elevator   
1                                Love, Hate, Love   
2                            Let Me Love You Baby   
3                                         My Love   
4    The Girl I Love She Got Long Black Wavy Hair   
..                                            ...   
109                                    Love Comes   
110                         Arms Around Your Love   
111                         Love Is a Losing Game   
112                         I Heard Love Is Blind   
113        (There Is) No Greater Love (Teo Licks)   

                                              composer  milliseconds  \
0                              Steven Tyler, Joe Perry        321828   
1                         Jerry Cantrell, Layne Staley        387134   
2                                         Willie Dixon        175386   
3                                     Jauperi/Zeu Góes        203493   
4    Jimmy Page/John Bonham/John Estes/John Paul Jo...        183327   
..                                                 ...           ...   
109  Darius "Take One" Minwalla/Jon Auer/Ken String...        199923   
110                                      Chris Cornell        214016   
111                                               None        154386   
112                                               None        129666   
113                          Isham Jones & Marty Symes        167933   

        bytes  unit_price   id  \
0    10552051        0.99    5   
1    12575396        0.99    7   
2     5716994        0.99   20   
3     6772813        0.99   29   
4     5995686        0.99   30   
..        ...         ...  ...   
109   3240609        0.99  265   
110   3516224        0.99  270   
111   2509409        0.99  321   
112   2190831        0.99  322   
113   2773507        0.99  322   

                                                title  artist_id  
0                                            Big Ones          3  
1                                            Facelift          5  
2    The Best Of Buddy Guy - The Millenium Collection         15  
3                                      Axé Bahia 2001         21  
4                        BBC Sessions [Disc 1] [Live]         22  
..                                                ...        ...  
109                               Every Kind of Light        200  
110                                          Carry On        205  
111                                     Back to Black        252  
112                                             Frank        252  
113                                             Frank        252  

[114 rows x 12 columns]

### Lister les top 5 artistes de Rock ayant le plus de morceaux.

In [88]:
import sqlite3
import pandas as pd

# Connect to the SQLite database
conn = sqlite3.connect('./data/jukebox.sqlite')

# Query the sqlite_master table to get the schema of all tables
query = '''
SELECT artists.name, COUNT (tracks.id) AS nb_tracks 
FROM genres
JOIN tracks ON genres.id = tracks.genre_id
JOIN albums ON albums.id = tracks.album_id
JOIN artists ON artists.id = albums.artist_id
WHERE genres.name = "Rock"
GROUP BY artists.name
ORDER BY nb_tracks DESC
LIMIT 5
'''

# Create a pandas DataFrame from the SQL query
df = pd.read_sql_query(query, conn)

# Display the DataFrame
df

name  nb_tracks
0  Led Zeppelin        114
1            U2        112
2   Deep Purple         92
3   Iron Maiden         81
4     Pearl Jam         54